### PR TITLE
Improve shape function of tf.sparse_concat when inputs are fully defined

### LINKED ISSUE #21964: tf.sparse_concat does not infer result shape from inputs with fully defined shape
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Windows 10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
NA
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
1.10
- **Python version**:
3.6.5
- **Bazel version (if compiling from source)**:
NA
- **GCC/Compiler version (if compiling from source)**:
NA
- **CUDA/cuDNN version**:
7.0
- **GPU model and memory**:
Yes
- **Exact command to reproduce**:
```python
In [9]: import tensorflow as tf

In [10]: x = tf.SparseTensor(indices=[[0,0],[1,1]], values=[1,2], dense_shape=[2,2])

In [11]: y = tf.SparseTensor(indices=[[0,0],[1,1]], values=[1,2], dense_shape=[2,2])

In [12]: x.get_shape()
Out[12]: TensorShape([Dimension(2), Dimension(2)])

In [13]: y.get_shape()
Out[13]: TensorShape([Dimension(2), Dimension(2)])

In [14]: z = tf.sparse_concat(-1, [x,y])

In [15]: z.get_shape()
Out[15]: TensorShape([Dimension(None), Dimension(None)])
```

The shape of z should be infered if shape of x and y are fully defined.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #21964 where the shape function of tf.sparse_concat always returns `TensorShape([Dimension(None), Dimension(None)])` even if the inputs are fully defined.

This fix addresses the issue by finding the shape if the inputs are fully defined.

This fix fixes #21964.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/python/ops/sparse_ops.py [hunk 1]
```diff
@@ -322,6 +322,12 @@ def sparse_concat(axis,
   output_ind, output_val, output_shape = (
       gen_sparse_ops.sparse_concat(inds, vals, shapes, axis, name=name))
 
+  shapes_value = [tensor_util.constant_value(shape) for shape in shapes]
+  if not shapes_value and all(shape is not None for shape in shapes_value):
+    dim = sum(shape[axis] for shape in shapes_value)
+    output_shape = shapes_value[0]
+    output_shape[axis] = dim
+    output_shape = tensor_shape.as_shape(output_shape)
```
Comment (alextp) [inline tensorflow/python/ops/sparse_ops.py]: I think output_shape is supposed to be a tensor and not a tensorshape

**Review by alextp — COMMENTED** [review]
Looks great! Just a minor nit

**Review by alextp — APPROVED** [review]

**Pull request merged**